src/landscape.cpp

```diff
@@ -751,48 +751,78 @@ std::tuple<CommandCost, Money> CmdClearArea(DoCommandFlag flags, TileIndex tile,
 }
 
 
-TileIndex _cur_tileloop_tile;
+/* Tile loop repeats pattern every MIN_MAP_SIZE tiles. Having it too small might
+ * be noticeable, And too big is a waste of resources. */
+static_assert(MIN_MAP_SIZE == 64);
 
-/**
- * Gradually iterate over all tiles on the map, calling their TileLoopProcs once every 256 ticks.
+static const uint TILE_LOOP_PATTERN_SIZE = MIN_MAP_SIZE * MIN_MAP_SIZE; // Size of the pattern
+static uint8_t _cur_tile_loop_pattern[TILE_LOOP_PATTERN_SIZE]; // Tile loop pattern, changes every 256 ticks
+
+uint32_t _cur_tileloop_seed;  // Seed value that was used to generate current tile loop pattern
+
+/*
+ * Determines whether the tile loop was already called for the tile in the
+ * current cycle (256 ticks).
+ * @param tile TileIndex of the tile
+ * @returns true iff tile was already iterated
  */
-void RunTileLoop()
+bool WasTileIteratedThisCycle(TileIndex tile)
 {
-	PerformanceAccumulator framerate(PFE_GL_LANDSCAPE);
+	uint pattern_index = (TileY(tile) % MIN_MAP_SIZE) * MIN_MAP_SIZE + TileX(tile) % MIN_MAP_SIZE;
+	return (_cur_tile_loop_pattern[pattern_index] <= TimerGameTick::counter % 256);
+}
 
+/*
+ * Generates a pattern to use for the tile loop seeded on the value of _cur_tileloop_seed
+ */
+void GenerateTileLoopPattern() {
 	/* The pseudorandom sequence of tiles is generated using a Galois linear feedback
 	 * shift register (LFSR). This allows a deterministic pseudorandom ordering, but
 	 * still with minimal state and fast iteration. */
 
-	/* Maximal length LFSR feedback terms, from 12-bit (for 64x64 maps) to 24-bit (for 4096x4096 maps).
-	 * Extracted from http://www.ece.cmu.edu/~koopman/lfsr/ */
-	static const uint32_t feedbacks[] = {
-		0xD8F, 0x1296, 0x2496, 0x4357, 0x8679, 0x1030E, 0x206CD, 0x403FE, 0x807B8, 0x1004B2, 0x2006A8, 0x4004B2, 0x800B87
-	};
-	static_assert(lengthof(feedbacks) == 2 * MAX_MAP_SIZE_BITS - 2 * MIN_MAP_SIZE_BITS + 1);
-	const uint32_t feedback = feedbacks[Map::LogX() + Map::LogY() - 2 * MIN_MAP_SIZE_BITS];
+	static int16_t lfsr_current = _cur_tileloop_seed % TILE_LOOP_PATTERN_SIZE;
+	const uint32_t LFSR_FEEDBACK = 0xD8F;  // 12-bit feedback term extracted from http://www.ece.cmu.edu/~koopman/lfsr/
+	static_assert(TILE_LOOP_PATTERN_SIZE == 4096);  // different size requires different feedback
```

Comment:
Unclear comment, is this what you mean?
```suggestion
	static_assert(TILE_LOOP_PATTERN_SIZE == 4096);  // If we ever allow larger maps, we need a different feedback term.
```